### ISSUE TITLE
Need some new operators and enchance some existing operators.

### ISSUE THREAD

**Issue by qingqing01** [PaddlePaddle/Paddle#12765]
### Need new ops:
- grid_sample :  https://pytorch.org/docs/stable/nn.html?highlight=grid_sample#torch.nn.functional.grid_sample
- affine_grid : https://pytorch.org/docs/stable/nn.html?highlight=affine_grid#torch.nn.functional.affine_grid
- psroi_pooling
- roi_align
- where

### Need to enhance：
- softmax:  do softmax on the specified axis
- gather, scatter:  index support more than one rank.

Welcome anyone to contribute.

**Comment by qingqing01** [comment]
- [x] Add grid_sample, affine_grid, psroi_pooling, roi_align
    Done
- [x] softmax: do softmax on the specified axis
    Done
